### ISSUE TITLE
Speed up painter's algorithm with bypass queue

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#774]
If we truly have thousands of planes, we're going to get hurt during rendering, no matter how small they all are, simply due to the iterated walk + geometry check. I.e., let's imagine we have 1000 planes plus the standard plane. All 1k are 1x1 planes at {2,2}. At every cell, we must perform 1000 pointer walks plus 1000 checks against plane geometries, all for nothing. This is unacceptable.

A viable solution could be keeping a data structure `punts[]` of O(n) slots for n planes. This is initialized to 0s. Keep a variable `firstpunt`, initialized to be N. The first cell, we check every plane. If the plane overlaps the cell, its value remains 0. Otherwise, *use the rectilinear nature of the display area to calculate the number of cells until we need check the plane again*. This is an O(1) calculation. Sort the result into `punts`. We know it's greater than any 0 punts, so do a binary insertion on the range `firstpunt..N` (primary index: calculation, secondary index: planar depth). Set `firstpunt` if we sort to `punts[s], s < firstpunt`. Easy-peasy.

At any given cell, we only need check those planes with no `punt`, in the order they appear in `punts`. We then need logically decrement all other punts by 1. This would be O(N), though still way better than cruising through all those planes (and destroying caches). Since renderings are strongly ordered, though, we can just keep a nonce, incremented by 1 on each render. When a punt is calculated, add the current nonce. When evaluating non-0 punts, subtract the current nonce. This operation is now O(1).

Finally, note that we only ever need to check from `firstpunt` through the first value which is greater than the nonce. So we perform exactly as many operations as we need, plus 1.

We're currently Θ(P*Y*X); with this change, we remain O(P*Y*X), but fall to Ω(Y*X) (**update: see correction below**). For large P with small planes (the only sensible case for large P), this is a huge improvement, and probably means the difference between success and failure, between victory and idle dreams.

mmmmmmmmm

**Comment by dankamongmen** [comment]
Actually I think it's Ω(Y*X*lgP), due to the sort, and actually O(YXP^2lgP) (in the true (phsycially unrealizeable) worst case of checkerboard planes having the full size. but if you're doing that, fuck you.

**Comment by dankamongmen** [comment]
Also we'd need to shuffle around some memory when we sort, unless we use something other than an array. A BST ought provide us both lgN index discovery and O(1) insertion, as opposed to lgN index discovery + O(N) move for the array.

**Comment by dankamongmen** [comment]
It goes without saying that for the case of few P, all this complexity is going to eat shit compared to the current simple solution. Maybe fall over to this only when P transcends some value? Or more precisely, when the average "plane density" falls below some value. But this will be absolutely necessary for thousands of planes.

I'm of course not sure who *wants* thousands of planes.

**Comment by dankamongmen** [comment]
If we wanted to avoid having to keep the expectation queue (`punts[]`) sorted, we could do something like only check against the smallest; if we're larger, go at the end. if we're smaller, replace the smallest. that's O(1). then when we popped the smallest, we'd need do an O(N) search for the next-smallest...no, that's likely worse; since there's one pop for each push, we've turned O(Nlgn) into O(n^2), whoops! don't do that.

**Comment by dankamongmen** [comment]
Note that if we do this, since we never move past a cell until it's solved (unlike now where we move from plane to plane *in toto*), we can eliminate `postpaint()` and its O(N) final loop over the viewing region. That ought be a nice win!

**Comment by dankamongmen** [comment]
> Note that if we do this, since we never move past a cell until it's solved (unlike now where we move from plane to plane _in toto_), we can eliminate `postpaint()` and its O(N) final loop over the viewing region. That ought be a nice win!

note that *this*, however, has been partially invalidated by moving to distinct render/raster--`postpaint()` has been moved into raster, since it needs compare against the most recently-written frame. of course, we could just to that while performing the raster walk. _potamoisi toisin autoisin embainousin hetera kai hetera hudata epirrei._

**Comment by dankamongmen** [comment]
I sat down today and worked this out in detail, and did some analysis.

**tl;dr: we can't beat the current scheme in the worst case, in terms of asymptotic analysis. it is possible that we could beat it in other ways with this new scheme, but not at all definite.**

i say in the initial writeup that we are Θ(PYX), which is loose: we only iterate over max(Y, PY)xmax(X, PX) for each plane, not YxX. i.e. if a plane is significantly smaller than the viewing area, we check significantly fewer cells.

that's not far from the compulsory number of cell checks. we absolutely have to check each cell of each plane until we find the solution for a cell. at that point, we might be able to cut some out, but we don't currently -- we currently always check exactly each cell of each plane exactly once. technically, we don't need to check any of the cells beneath a cell solution.

we achieve that with the new algorithm -- we never go deeper than we need to, on a per-cell basis. that's guaranteed. there are no guaranteed savings from this, though. take the trivial case of a single, standard plane. each cell solves its cell. each cell must be checked.

so any advantages must come from
* not examining cells below a solution
* better cache behavior
* eliminating some second pass

i mentioned the chance of inlining `postpaint()` above, but that doesn't work anymore, since we've split up render and raster. we don't know the lastframe at render time, and thus can't compute damage, and thus can't inline `postpaint()` into rendering (see #1262 about inducting it into raster, though). We can inline `rvec_init()`, though probably not to great effect.

An 80x70 frame of 16B cells is ~89K, so cache behavior is important. By doing the new way, we'd eliminate any duplicate trips of our solution plane through a small L1 cache. Behavior of the planes is difficult to predict. Our working set for each chunk of iterations is small enough to fit within L1, so a lot of this would come down to how well prefetching can hide cache misses in the current solution.

The new solution ought have a great many fewer branch mispredictions by virtue of having many, many fewer branches overall.

I improved the algorithm as initially described, eliminating all but the initial sort via clever exploitation of the rectilinearity of it all.

The rendering area is RY * RX, where RY and RX are positive integers.

A plane is either active or inactive for a given cell in the rendering area.
The plane is active if it is defined at that cell. It is inactive otherwise.

There is an initial (possibly empty) inactive region before the plane is first
reached. There then follow A' (A' >= 0) active regions, separated by
(I' = A'-1) inactive regions (I' is 0 if A' is 0). These active regions A_0,
A_1, ... all have the same size, and these inactive regions I_0, I_1, ...
likewise all have the same size. I_0 + A_0 == RX. There is then a final
(possibly empty) inactive region following the plane's lowermost, rightmost
intersection with the visual area.

 I_init + A' * A_0 + I' * I_0 + I_final == RX * RY.

Given RX and RY, we can describe a plane's activity pattern completely with
three numbers: I_init, A', and A_0. 

Keep two ordered structures, an active set and an inactive set. The active set
is counting down until they become inactive. The inactive set is counting down
until the become active.

Initialization:

For each plane, calculate I_init and A_0. Planes with I_init values of 0 go
into the active set, sorted first by A_0 and secondarily by plane depth. Planes
with I_init values >= 0 go into the inactive set, sorted first by I_init and
secondarily by plane depth.

For rendering area RY * RX and plane py * px at offset y, x, I_init is:

 infinite for x >= RX
 infinite for x + px <= 0
 infinite for y >= RY
 infinite for y + py <= 0 
 0 for y <= 0, y + py >= 0, x <= 0
 x for y <= 0, y + py >= 0, x > 0
 y * RX + x for y > 0, x >= 0
 y * RX for y > 0

max finite initial gap is RY * RX - 1. min initial gap is 0.

Each node is a pointer to a plane, and the scalar coordinate xy
(0 <= xy < PX * PY) at which the current state changes (A_0 and I_init).

assuming finite initial gap (i.e. that the plane overlaps the rendering area),
the active length (can exceed practical length) is:

 x <= 0:
  x + px >= RX: (spans horizontal range)
   y <= 0:
     RX * py + y, from origin
   y > 0:
     RX * py, from column 0
  x + px < RX:
    x + px, 
 x > 0:
  x + px >= RX:
    RX - x
  x + px < RX:
    px

max active length is RY * RX (for a plane covering the entirety of the
horizontal viewing area), otherwise RX - 1. min active length is 1.

inactive gap is undefined if plane spans visual region or is invisible.
otherwise, inactive gap is calculated at right edge of plane (column C),
and is equal to PX - (C + 1) + x if x >= 0, or PX - (C + 1) otherwise.

at each step we check to see if the foremost planes of either set need flip
to the other set. this suggests an extra sort per flip. unless we've eclipsed
a plane's I_init, or entered a plane's I_final, an element moving from one set
to another must have the same previous element as it did before. each node
thus keeps an additional element, a double pointer to the previous element's
next link. upon flip, check this pointer to ensure it's NULL. if it is NULL,
link ourselves. otherwise, chase to the end, and link ourselves.

**Comment by dankamongmen** [comment]
It is with a heavy heart that I'm closing this as NONFIX, perhaps to one day be reopened. It's a substantial and delicate bit of change at our deepest layers, and profiles (see #1262) just don't identify this as a serious target of opportunity. We'd have to be very lucky to see a big win off of it anyway, at least in any kind of typical usage case.

At least half or more of `paint()` is in channel blending.

With that said, channel blending could just be representing the cache load of `vis`, if `crender->p != NULL`...in which case the new algorithm could improve our performance.

I'm just not seeing it. Why is rasterization more expensive than rendering (they're about equal when leaving out the actual writeout)? *That's* a question worth answering. If raster is slower than render despite only having to go throuhg one assembled plane, I doubt speeding up plane traversal (especially at the cost of high code complexity) is that big a win.

Still worth trying one day, maybe. But not now.
